src/Adapter/MSTest.TestAdapter/Execution/TypeCache.cs

```diff
@@ -260,24 +260,42 @@ private static bool TryGetUnescapedManagedTypeName(TestMethod testMethod, [NotNu
     private TestClassInfo CreateClassInfo(Type classType, TestMethod testMethod)
     {
         IEnumerable<ConstructorInfo> constructors = PlatformServiceProvider.Instance.ReflectionOperations.GetDeclaredConstructors(classType);
-        ConstructorInfo? constructor = constructors.FirstOrDefault(ctor => ctor.GetParameters().Length == 0 && ctor.IsPublic);
-        bool isParameterLessConstructor;
+        (ConstructorInfo? CtorInfo, bool IsParameterless)? selectedConstructor = null;
 
-        if (classType.GetConstructor([typeof(TestContext)]) is { } testContextCtor)
+        foreach (ConstructorInfo ctor in constructors)
         {
-            constructor = testContextCtor;
-            isParameterLessConstructor = false;
-        }
-        else if (classType.GetConstructor([]) is { } parameterLessCtor)
-        {
-            constructor = parameterLessCtor;
-            isParameterLessConstructor = true;
+            if (!ctor.IsPublic)
+            {
+                continue;
+            }
+
+            ParameterInfo[] parameters = ctor.GetParameters();
+
+            // There are just 2 ctor shapes that we know, so the code is quite simple,
+            // but if we add more, add a priority to the search, and shortcircuit this search so we only iterate
+            // through the collection once, to avoid re-allocating GetParameters multiple times.
+            if (parameters.Length == 1 && parameters[0].ParameterType == typeof(TestContext))
+            {
+                selectedConstructor ??= (ctor, IsParameterless: false);
+
+                // This is the preferred constructor, no point in searching for more.
+                break;
+            }
+
+            if (parameters.Length == 0)
+            {
+                selectedConstructor = (ctor, IsParameterless: true);
```

Comment:
Maybe add `Debug.Assert(selectorConstructor is null);` before this line?

Reply:
moving the `??=` to the correct place solves the issue that was motivation for this Assert. Which was to error when we forget to break in the most preferred case, then falling through here would choose the non-preferred constructor.